File: practice/advanced contest/AVGMAT/Avgmat.java
import java.util.*;
import java.lang.*;

public class Avgmat{
    private static int manhattanDistance(int[] a, int[] b){
        // distance is just the sum of differences of axes
        return Math.abs(a[0] - b[0]) + Math.abs(a[1] - b[1]);
    }
    private static void runTestCase(Scanner sc){
        List<int[]> cellsList = new LinkedList<>(); // need a set of integer pairs for coordinates
        // each 'first line' in test case is two space separated integers, n and m
        String[] nm = sc.nextLine().split(" ");
        int n = Integer.parseInt(nm[0]);
        int m = Integer.parseInt(nm[1]);
        for (int i = 0; i<n; i++){
            char[] line = sc.nextLine().toCharArray();
            for (int j = 0; j < line.length; j++){
                // each character is either a 0 or 1; if 1, then it is a house
                if (line[j] == '1'){
                    int[] cell = {i, j};
                    cellsList.add(cell);
                }
            }
        }

        // at this point, we have the set of all List in this test case
        // get the distance of all cells; add them to a map from distance -> cell
        int[] distanceCount = new int[n+m-2];
        int[][] cells = new int[cellsList.size()][2];
        cellsList.toArray(cells);
        for (int i=0; i<cells.length; i++){
            for (int j=i+1; j<cells.length; j++){
                distanceCount[manhattanDistance(cells[i], cells[j]) - 1]++;
            }
        }
        for (int i=0; i<distanceCount.length; i++){
            System.out.print(distanceCount[i] + " ");
        }
        System.out.println();
    }
    public static void main(String[] args) throws java.lang.Exception{
        Scanner sc = new Scanner(System.in);
        int t = Integer.parseInt(sc.nextLine());    // first line is always number of test cases
        for (int i=0; i<t; i++)
            runTestCase(sc);
    }
}

File: practice/advanced contest/AVGMAT/avgmat.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/**
 * each test case is a grid n*m, each cell is either empty or contains a house
 * distance between houses is the manhattan distance between the cells containing them
 * for each d between 1 and (N + M - 2) inclusive, chef wants to calculate the number of unordered pairs of distinct houses with distance equal to d
 * 
 * @return int 
 */

int abs(int x){
    return (x < 0) ? -x : x;
}

int manhattandistance(int a[], int b[]){
    return abs(a[0] - b[0]) + abs(a[1] - b[1]);
}

int main(void)
{

    int t; // number of test cases
    scanf("%d", &t);

    for (int c = 0; c < t; c++)
    {
        // each case has two integers up front: n, m
        int n, m;
        scanf("%d %d", &n, &m);
        int cells[n * m][2]; // coordinate sets
        // memset(cells, 0, n * m); // 0 out the cells
        int numcells = 0;

        // n is the number of lines that follow. Injest these lines:
        for (int i = 0; i < n; i++)
            for (int j = 0; j < m; j++)
            {
                int temp;
                scanf("%1d", &temp);
                if (temp == 1)
                {
                    cells[numcells][0] = i;
                    cells[numcells][1] = j;
                    ++numcells;
                }
            }
        int distmax = n + m - 2;
        int distances[distmax];
        memset(distances, 0, sizeof(distances));
        for (int i=0; i<numcells; i++){
            for (int j=i+1; j<numcells; j++){
                ++distances[manhattandistance(cells[i], cells[j]) - 1];
            }
        }
        for (int i=0; i<distmax; i++){
            printf("%d ", distances[i]);
        }
        printf("\n");
    }
    exit(0);
}